src/EditorFeatures/Core/Implementation/Outlining/OutliningTaggerProvider.cs

```diff
@@ -44,18 +53,32 @@ internal partial class OutliningTaggerProvider :
             IEditorOptionsFactoryService editorOptionsFactoryService,
             IProjectionBufferFactoryService projectionBufferFactoryService,
             [ImportMany] IEnumerable<Lazy<IAsynchronousOperationListener, FeatureMetadata>> asyncListeners)
-            : base(new AggregateAsynchronousOperationListener(asyncListeners, FeatureAttribute.Outlining), notificationService)
+                : base(new AggregateAsynchronousOperationListener(asyncListeners, FeatureAttribute.Outlining), notificationService)
         {
             _textEditorFactoryService = textEditorFactoryService;
             _editorOptionsFactoryService = editorOptionsFactoryService;
             _projectionBufferFactoryService = projectionBufferFactoryService;
         }
 
-        protected override bool RemoveTagsThatIntersectEdits => true;
+        public override IEqualityComparer<IOutliningRegionTag> TagComparer => this;
 
-        protected override SpanTrackingMode SpanTrackingMode => SpanTrackingMode.EdgeExclusive;
+        bool IEqualityComparer<IOutliningRegionTag>.Equals(IOutliningRegionTag x, IOutliningRegionTag y)
+        {
+            // This is only called if the spans for the tags were the same. In that case, we consider ourselves the same
+            // unless the CollapsedForm properties are different.
+            return object.Equals(x.CollapsedForm, y.CollapsedForm);
+        }
+
+        int IEqualityComparer<IOutliningRegionTag>.GetHashCode(IOutliningRegionTag obj)
+        {
+            // This will not result in lots of hash collisions as our caller will
+            // first be hashing spans, and then adding this value to that.
+            // The only collisions will be for outlining tags with the same span
+            // (which is what we want).
+            return 0;
```

Comment:
Change for another pull request: any reason not to just call object.GetHashCode(obj.CollapsedForm)? might be shorter than the comment explaining why we're not doing that.